### ISSUE TITLE
Language service not loading with config option "files: []"

### ISSUE THREAD

**Issue by lincolnthree** [angular/vscode-ng-language-service#793]
**Describe the bug**

Angular HTML Template syntax validation/completion/intelli-sense are not working.

**To Reproduce**

1. `npm install -g @angular/cli@next`
2. `ng new angular10example`
3. `cd angular10example`
4. `npm i`
5. `Open project in VSCode with Angular Language Service extension installed`

**Expected behavior**

Angular HTML template editor should provide validition, syntax checking, completion and "go to definition" functionality. Etc etc.

**Logs**

Please attach two logs:

1. Console output (Go to toolbar --> View --> Output, and copy the contents)

````
[Info  - 4:28:28 PM] Angular language server process ID: 34506
[Info  - 4:28:28 PM] Using typescript v3.8.3 from /Users/lincoln/.vscode/extensions/angular.ng-template-0.901.9/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 4:28:28 PM] Using @angular/language-service v9.1.9 from /Users/lincoln/.vscode/extensions/angular.ng-template-0.901.9/server/node_modules/@angular/language-service/bundles/language-service.umd.js
[Info  - 4:28:28 PM] Log file: /Users/lincoln/Library/Application Support/Code/logs/20200526T132525/exthost4/Angular.ng-template/nglangsvc.log
[Info  - 4:28:30 PM] Disabling language service for /Users/lincoln/projects/maritlabs/libraries/angular10example/tsconfig.json because it is not an Angular project ('@angular/core/core.d.ts' could not be found). If you believe you are seeing this message in error, please reinstall the packages in your package.json.
[Error - 4:28:30 PM] No config file for /Users/lincoln/projects/maritlabs/libraries/angular10example/src/app/app.component.html
````

2. Log file (Path is printed to the console output at startup)

[nglangsvc.log](https://github.com/angular/vscode-ng-language-service/files/4684891/nglangsvc.log)

**Screenshots**

![image](https://user-images.githubusercontent.com/362329/82946888-22a96580-9f6d-11ea-818d-c759c74f88cf.png)

![image](https://user-images.githubusercontent.com/362329/82947393-f9d5a000-9f6d-11ea-8af7-175cb4248a13.png)

**Additional context**

Not sure when this stopped working. It actually hasn't been working for me for about a week, even in my Angular 9 project, and I haven't been able to figure out what's causing the problem, but I've since upgraded to Angular 10 and it's still not working.

**Comment by lincolnthree** [comment]
I have also tried using TypeScript 3.9.3 from the project TS version.

**Comment by lincolnthree** [comment]
I have tried setting `angular.nsdk` to use the `@angular/language-service` version 10.0.0-rc1 from the project's node_modules.

**Comment by lincolnthree** [comment]
From another workspace:

````
[Info  - 4:35:45 PM] Angular language server process ID: 34938
[Info  - 4:35:45 PM] Using typescript v3.9.3 from /Users/lincoln/projects/maritlabs/topdecked-unified/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 4:35:45 PM] Using @angular/language-service v10.0.0-rc.0 from /Users/lincoln/projects/maritlabs/topdecked-unified/node_modules/@angular/language-service/bundles/language-service.umd.js
[Info  - 4:35:45 PM] Log file: /Users/lincoln/Library/Application Support/Code/logs/20200526T132525/exthost1/Angular.ng-template/nglangsvc.log
[Info  - 4:35:51 PM] Disabling language service for /Users/lincoln/projects/maritlabs/topdecked-unified/tsconfig.json because it is not an Angular project ('@angular/core/core.d.ts' could not be found). If you believe you are seeing this message in error, please reinstall the packages in your package.json.
[Info  - 4:35:51 PM] Enabling language service for /Users/lincoln/projects/maritlabs/topdecked-unified/tsconfig.app.json.
[Info  - 4:35:51 PM] Enabling language service for /Users/lincoln/projects/maritlabs/topdecked-unified/tsconfig.worker.json.
[Info  - 4:35:51 PM] Disabling language service for /Users/lincoln/projects/maritlabs/topdecked-unified/e2e/tsconfig.e2e.json because it is not an Angular project ('@angular/core/core.d.ts' could not be found). If you believe you are seeing this message in error, please reinstall the packages in your package.json.
[Error - 4:35:51 PM] No config file for /Users/lincoln/projects/maritlabs/topdecked-unified/src/app/pages/home/home.page.htm
````

**Comment by ayazhafiz** [comment]
Thanks for the report!

I could not reproduce this with the steps you provided. Could you verify that you've opened VSCode to the root of the project, and the root contains `node_modules/@angular/core/core.d.ts`?

**Comment by lincolnthree** [comment]
@ayazhafiz You're welcome! Thanks for taking a look.

Hoping I can figure out what's going on here. Been burning a lot of time without these features :/

```
sharktop:topdecked-unified lincoln$ ls node_modules/@angular/core/core.d.ts
-rw-r--r--  1 lincoln  staff  1258907 May 26 16:20 node_modules/@angular/core/core.d.ts
sharktop:topdecked-unified lincoln$
```

Same behavior. When run from project directory:

Ran VSCode:
```
sharktop:topdecked-unified lincoln$ code .
```

Opened app.component.html, VSCode NGLS Log:
```
[Info  - 9:04:01 AM] Angular language server process ID: 37826
[Info  - 9:04:01 AM] Using typescript v3.8.3 from /Users/lincoln/.vscode/extensions/angular.ng-template-0.901.9/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 9:04:01 AM] Using @angular/language-service v10.0.0-rc.0 from /Users/lincoln/projects/maritlabs/topdecked-unified/node_modules/@angular/language-service/bundles/language-service.umd.js
[Info  - 9:04:01 AM] Log file: /Users/lincoln/Library/Application Support/Code/logs/20200527T090334/exthost1/Angular.ng-template/nglangsvc.log
[Info  - 9:04:03 AM] Disabling language service for /Users/lincoln/projects/maritlabs/topdecked-unified/tsconfig.json because it is not an Angular project ('@angular/core/core.d.ts' could not be found). If you believe you are seeing this message in error, please reinstall the packages in your package.json.
[Error - 9:04:03 AM] No config file for /Users/lincoln/projects/maritlabs/topdecked-unified/src/app/app.component.html
```

Completion / etc still not working :/

**Comment by lincolnthree** [comment]
These are my other installed extensions, for reference:

![image](https://user-images.githubusercontent.com/362329/83022564-b4f74b00-9ff9-11ea-9165-bbc6363fa145.png)

**Comment by lincolnthree** [comment]
I also should have reported that this is occurring on MacOS.

**Comment by lincolnthree** [comment]
What other info should I be looking at / what can I provide to get to the bottom of this?

**Comment by lincolnthree** [comment]
My uneducated overview of the log file leads me to believe that there is an issue with the new reference-based tsconfig.json structure in Angular 10 projects causing the language service tofail to resolve typescript and other various other file roots and settings. But that doesn't hold up since you said it worked for you.

Log from the last attempt:

[nglangsvc.log](https://github.com/angular/vscode-ng-language-service/files/4689044/nglangsvc.log)

**Comment by lincolnthree** [comment]
I've been 'hacking' tsconfig.json a bit to try to get the angular language service to recognize it as an angular project... finally ended up with something a little different:
```
{
  "extends": "../tsconfig.base.json",
  "files": [],
  "references": [{
      "path": "./tsconfig.app.json"
    },
    {
      "path": "./tsconfig.spec.json"
    },
    {
      "path": "./tsconfig.worker.json"
    }
  ],
  "include": [
    "node_modules"
  ],
  "angularCompilerOptions": {
    "enableIvy": true,
    "fullTemplateTypeCheck": false,
    "skipTemplateCodegen": false,
    "skipTemplates": true,
    "strictInjectionParameters": true,
    "strictMetadataEmit": false
  }
}
```


```
[Info  - 9:43:15 AM] Angular language server process ID: 39364
[Info  - 9:43:15 AM] Using typescript v3.9.3 from /Users/lincoln/projects/maritlabs/topdecked-unified/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 9:43:15 AM] Using @angular/language-service v9.1.9 from /Users/lincoln/.vscode/extensions/angular.ng-template-0.901.9/server/node_modules/@angular/language-service/bundles/language-service.umd.js
[Info  - 9:43:15 AM] Log file: /Users/lincoln/Library/Application Support/Code/logs/20200527T093711/exthost1/Angular.ng-template/nglangsvc.log
[Info  - 9:43:32 AM] Enabling language service for /Users/lincoln/projects/maritlabs/topdecked-unified/tsconfig.json.
[Error - 9:43:32 AM] No config file for /Users/lincoln/projects/maritlabs/topdecked-unified/src/app/pages/home/home.page.html
Warning: Can't resolve all parameters for Router in /Users/lincoln/projects/maritlabs/topdecked-unified/node_modules/@angular/router/router.d.ts: ([object Object], [object Object], [object Object], [object Object], [object Object], [object Object], [object Object], ?). This will become an error in Angular v6.x
Warning: Can't resolve all parameters for ɵangular_packages_router_router_o in /Users/lincoln/projects/maritlabs/topdecked-unified/node_modules/@angular/router/router.d.ts: ([object Object], [object Object], ?). This will become an error in Angular v6.x

```

[nglangsvc.log](https://github.com/angular/vscode-ng-language-service/files/4689145/nglangsvc.log)

**Comment by ayazhafiz** [comment]
What did your tsconfig look like originally?

**Comment by ayazhafiz** [comment]
The one listed in the log file is the output after the config file is parsed by TS, so it doesn't always line up (e.g. in regard to rootNames)

**Comment by ayazhafiz** [comment]
The issue is that `files: []` tells TypeScript's tsserver to load a project with no files (other than the config file). This project is passed to the Angular language service. The project is also all the language service has to work with, so if core.d.ts or any TypeScript files can't be found in the project files (as in this case), the language service is disabled.

**Comment by lincolnthree** [comment]
The original is the TS config file that was generated by angular when I created the example project in the reproduction steps listed in the original issue report.

I will post it when I get back to my computer, but I believe you should have it on your computer if you generated the example.

**Comment by lincolnthree** [comment]
Original tsconfig.json, as generated:

```
// This is a "Solution Style" tsconfig.json file, and is used by editors and TypeScript’s language server to improve development experience.
// It is not intended to be used to perform a compilation.
{
  "files": [],
  "references": [
    {
      "path": "./tsconfig.app.json"
    },
    {
      "path": "./tsconfig.spec.json"
    },
    {
      "path": "./e2e/tsconfig.json"
    }
]
}

```

**Comment by lincolnthree** [comment]
I guess it's worth noting that the Angular language service does resolve the referenced tsconfig.app.json/etc projects, and does enable language services for those... then subsequently fails to resolve configuration for any HTML files.

**Comment by lincolnthree** [comment]
That being said, I deleted `files: []` from the generated tsconfig.json file and the language service has started functioning in html files again. (If it weren't clear, `ng new` is what generated this file (as did the `ng update` that I ran in my main project.

Interesting that it's generating a tsconfig file that seems to break the language service, but... you said it worked for you. 

Out of curiosity, does your generated `tsconfig.json` file match what I pasted?

**Comment by ayazhafiz** [comment]
Failing to find a config for HTML files isn't a big deal, since the config files are only relevant for TS sources.

The relevant validation code:

https://github.com/angular/vscode-ng-language-service/blob/da08740584d607f340a57fbd5b92dc4319874655/server/src/session.ts#L475-L486

I think we can fall back on trying to resolve node_modules if the project files are empty (not sure of all the implications of this), but for now removing `files: []` is a workaround.

**Comment by lincolnthree** [comment]
(Also thank you for looking into this. At this point my environment is functional, so if you choose to call it resolved, that's fine, but I suspect that others would be running into this as well if my assumptions are correct -- and I'm happy to keep digging.)

**Comment by ayazhafiz** [comment]
> Out of curiosity, does your generated `tsconfig.json` file match what I pasted?

It didn't; I probably installed the cli locally when I tried this yesterday, but did it again an hour ago and got the same config you did.

**Comment by ayazhafiz** [comment]
Yeah, let's leave this open for now. We don't want the only solution to delete a config auto-generated by Angular.

**Comment by lincolnthree** [comment]
Awesome. Glad I could be of assistance! Let me know if there's anything I can do to help.

**Comment by lincolnthree** [comment]
Thank you again.

**Comment by lincolnthree** [comment]
Uh oh. 

One issue I just encountered as a repercussion of deleting `"files": []`... vscode typescript stops properly resolving source files:

![image](https://user-images.githubusercontent.com/362329/83043361-c730b380-a010-11ea-8ccf-c5ba258164de.png)

And this issue:
![image](https://user-images.githubusercontent.com/362329/83043411-dca5dd80-a010-11ea-9f2e-7715e797fa6a.png)

So it seems that this setting is required to get the typescript language service to function properly.

**Comment by lincolnthree** [comment]
So we can either have typescript or functional angular :/

**Comment by lincolnthree** [comment]
Adding the following to files: will at least get things working for the app sources, but I don't know what kind of other impacts this will have:

```
  "files": [
    "src/main.ts",
    "src/polyfills.ts",
    "src/zone-flags.ts"
  ]
```

**Comment by ayazhafiz** [comment]
If anyone wants to contribute a patch to fix this, I think the best way would be to replace the existing Angular project check (https://github.com/angular/vscode-ng-language-service/issues/793#issuecomment-634727597) with an activation event (https://github.com/angular/vscode-ng-language-service/issues/481). One thing to check is that the language service still behaves properly (i.e. just provides TS completions) if a TS-only subfolder is opened in a workspace.

**Comment by c0dybrown** [comment]
I might spend some more time on this later, but for anyone who doesn't care about performance or good practices and just wants to use this extension's features (don't thumbs me down for this):

1. Open this file:
`C:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js`

2. Find and replace this with an empty string:
`if (e_1) throw e_1.error;`

Everything will work, but the extension will never properly initialize so you will see an infinite initializing message 

Here is the actual error, something is going on with the new TS structure:

```
Warning: Can't resolve all parameters for Router in c:/Users/<user>/source/repos/<project>/<project>/ClientApp/node_modules/@angular/router/router.d.ts: ([object Object], [object Object], [object Object], [object Object], [object Object], [object Object], [object Object], ?). This will become an error in Angular v6.x
Warning: Can't resolve all parameters for ɵangular_packages_router_router_o in c:/Users/<user>/source/repos/<project>/<project>/ClientApp/node_modules/@angular/router/router.d.ts: ([object Object], [object Object], ?). This will become an error in Angular v6.x
c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:30390
                finally { if (e_1) throw e_1.error; }
                                   ^

TypeError: Cannot read property 'kind' of undefined
    at getConditionalFlowTypeOfType (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:45510:28)
    at getTypeFromTypeNode (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:47396:20)
    at getTypeOfNode (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:67279:40)
    at Object.getTypeAtLocation (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:35196:31)
    at getTsTypeFromBuiltinType (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:28433:28)
    at TypeScriptSymbolQuery.getBuiltinType (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:27600:28)
    at AstType.visitLiteralPrimitive (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:26298:47)
    at LiteralPrimitive.visit (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:7386:28)
    at ASTWithSource.visit (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:7542:29)
    at AstType.getDiagnostics (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:26031:28)
```

**Comment by c0dybrown** [comment]
I might spend some more time on this later, but for anyone who doesn't care about performance or good practices and just wants to use this extension's features (don't thumbs me down for this):

1. Open this file:
`C:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js`

2. Find and replace this with an empty string:
`if (e_1) throw e_1.error;`

Everything will work, but the extension will never properly initialize so you will see an infinite initializing message 

Here is the actual error, something is going on with the new structure:

```
Warning: Can't resolve all parameters for Router in c:/Users/<user>/source/repos/<project>/<project>/ClientApp/node_modules/@angular/router/router.d.ts: ([object Object], [object Object], [object Object], [object Object], [object Object], [object Object], [object Object], ?). This will become an error in Angular v6.x
Warning: Can't resolve all parameters for ɵangular_packages_router_router_o in c:/Users/<user>/source/repos/<project>/<project>/ClientApp/node_modules/@angular/router/router.d.ts: ([object Object], [object Object], ?). This will become an error in Angular v6.x
c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:30390
                finally { if (e_1) throw e_1.error; }
                                   ^

TypeError: Cannot read property 'kind' of undefined
    at getConditionalFlowTypeOfType (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:45510:28)
    at getTypeFromTypeNode (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:47396:20)
    at getTypeOfNode (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:67279:40)
    at Object.getTypeAtLocation (c:\Users\<user>\source\repos\<project>\<project>\ClientApp\node_modules\typescript\lib\tsserverlibrary.js:35196:31)
    at getTsTypeFromBuiltinType (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:28433:28)
    at TypeScriptSymbolQuery.getBuiltinType (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:27600:28)
    at AstType.visitLiteralPrimitive (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:26298:47)
    at LiteralPrimitive.visit (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:7386:28)
    at ASTWithSource.visit (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:7542:29)
    at AstType.getDiagnostics (c:\Users\<user>\.vscode\extensions\angular.ng-template-0.901.9\server\node_modules\@angular\language-service\bundles\language-service.umd.js:26031:28)
```

**Comment by ayazhafiz** [comment]
@c0dybrown this sounds like a separate issue. Please see if #797 solves it; if not, feel free to open a new issue.

**Comment by kyliau** [comment]
tldr; Angular extension v0.1000.0-rc.0 has been released and can be downloaded [here](https://github.com/angular/vscode-ng-language-service/releases/tag/v0.1000.0-rc.0). If you'd like to try it out, please keep in mind that it has to be installed manually. See [here](https://github.com/angular/vscode-ng-language-service#installing-a-particular-release-build) for installation instructions.

- typescript 3.9 introduced "[solution-style](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#support-for-solution-style-tsconfigjson-files)" tsconfig, and it is not supported by prior ts versions.
- `latest` version of Angular (v9) and the language service extension uses typescript 3.8.
- "solution-style" tsconfig is enabled in Angular CLI in https://github.com/angular/angular-cli/pull/17586.
- In order for this to work, language service extension has to upgrade to both angular v10 and typescript 3.9. We've just released a rc version of the extension that supports this. Since this is rc, the binary is not deployed to visual studio marketplace, so installation has to be done manually.

Hopefully this clears up the confusion why the extension is not working for users who are currently using Angular v10 rc. As always, please let us know if you run into any issues, and thank you for helping us test v10 rc!

**Comment by kyliau** [comment]
I'm closing this issue in favor of #824.
